crates/bevy_tasks/src/bevy_executor.rs

```diff
@@ -0,0 +1,1296 @@
+use concurrent_queue::ConcurrentQueue;
+use futures_lite::{future,FutureExt};
+use slab::Slab;
+use thread_local::ThreadLocal;
+use crossbeam_utils::CachePadded;
+
+// ThreadLocalState *must* stay `Sync` due to a currently existing soundness hole.
+// See: https://github.com/Amanieu/thread_local-rs/issues/75
+static THREAD_LOCAL_STATE: ThreadLocal<ThreadLocalState> = ThreadLocal::new();
+
+pub(crate) fn install_runtime_into_current_thread(executor: &Executor) {
+    // Use LOCAL_QUEUE here to set the thread destructor
+    LOCAL_QUEUE.with(|_| {
+        let tls = THREAD_LOCAL_STATE.get_or_default();
+        let state = executor.state_as_arc();
+        let state_ptr = Arc::into_raw(state).cast_mut();
+        let old_ptr = tls.executor.swap(state_ptr, Ordering::Relaxed);
+        if !old_ptr.is_null() {
+            // SAFETY: If this pointer was not null, it was initialized from Arc::into_raw.
+            drop(unsafe { Arc::from_raw(old_ptr) });
+        }
+    });
+}
+
+std::thread_local! {
+    static LOCAL_QUEUE: CachePadded<UnsafeCell<LocalQueue>> = const {
+        CachePadded::new(UnsafeCell::new(LocalQueue {
+            local_queue: VecDeque::new(),
+            local_active: Slab::new(),
+        }))
+    };
+}
+
+/// # Safety
+/// This must not be accessed at the same time as `LOCAL_QUEUE` in any way.
+#[inline(always)]
+unsafe fn try_with_local_queue<T>(f: impl FnOnce(&mut LocalQueue) -> T) -> Result<T, AccessError> {
+    LOCAL_QUEUE.try_with(|tls| {
+        // SAFETY: This value is in thread local storage and thus can only be accessed
+        // from one thread. The caller guarantees that this function is not used with
+        // LOCAL_QUEUE in any way.
+        f(unsafe { &mut *tls.get() })
+    })
+}
+
+struct LocalQueue {
+    local_queue: VecDeque<Runnable>,
+    local_active: Slab<Waker>,
+}
+
+impl Drop for LocalQueue {
+    fn drop(&mut self) {
+        for waker in self.local_active.drain() {
+            waker.wake();
+        }
+
+        while self.local_queue.pop_front().is_some() {}
+    }
+}
+
+struct ThreadLocalState {
+    executor: AtomicPtr<State>,
+    stealable_queue: ConcurrentQueue<Runnable>,
+    thread_locked_queue: ConcurrentQueue<Runnable>,
+}
+
+impl Default for ThreadLocalState {
+    fn default() -> Self {
+        Self {
+            executor: AtomicPtr::new(core::ptr::null_mut()),
+            stealable_queue: ConcurrentQueue::bounded(512),
+            thread_locked_queue: ConcurrentQueue::unbounded(),
+        }
+    }
+}
+
+/// A task spawner for a specific thread. Must be created by calling [`TaskPool::current_thread_spawner`]
+/// from the target thread.
+///
+/// [`TaskPool::current_thread_spawner`]: crate::TaskPool::current_thread_spawner
+#[derive(Clone, Debug)]
+pub struct ThreadSpawner<'a> {
+    thread_id: ThreadId,
+    target_queue: &'static ConcurrentQueue<Runnable>,
+    state: Arc<State>,
+    _marker: PhantomData<&'a ()>,
+}
```

Comment:
As a nit, I found this name slightly confusing. It sounds like it spawns threads, not "a spanwer local to the thread".

Reply:
I went ahead and renamed it `LocalTaskSpawner` and `MainThreadTaskSpawner` instead.

Reply:
That's much clearer to me.